### PR TITLE
Add support for maplike construct in WebIDL contiguous

### PR DESCRIPTION
**Pull request by dontcallmedom**

### CODE REVIEW

**Comment by dontcallmedom** [comment]
Ping? This and #534 are needed to make it possible to migrate the WebRTC spec to contiguous mode.

**Comment by halindrome** [comment]
Unfortunately I don't know anything about WebIDL so I can't really evaluate - sorry!  @tobie?

File: js/webidl2.js [hunk 1]
```diff
@@ -582,7 +582,7 @@
                 return ret;
             }
             else if (consume(ID, "stringifier")) {
-                ret.stringifier = true;
+                ret.stringifier = true;-
```
Comment (tobie) [inline js/webidl2.js]: O_o

Comment (dontcallmedom) [inline js/webidl2.js]: this is a copy of webidl2.js from the its repo; I suggest NOT to review changes in this file

(ideally, we would not copy but import webidl2.js; but that's another story)

Comment (tobie) [inline js/webidl2.js]: I know. It Just jumped at me. Will look at the rest.

File: js/core/webidl-contiguous.js [hunk 2]
```diff
@@ -656,6 +671,9 @@ define(
                                           return optional + idlType2Text(arg.idlType).toLowerCase() + variadic;
                                       }).join(',').replace(/\s/g, '_') + ')');
                         break;
+                    case "maplike":
+                        name = "maplike";
+                        defn.idlId = "idl-def-" + parent.toLowerCase() + "-" + name.toLowerCase();
```
Comment (marcoscaceres) [inline js/core/webidl-contiguous.js]: Nit: 
("idl-def-" + parent + "-" + name).toLowerCase()

File: js/webidl2.js (hunk 1 above)
Comment (marcoscaceres) [inline js/webidl2.js]: I had the same reaction.

File: js/webidl2.js [hunk 3]
```diff
@@ -687,6 +687,69 @@
...
+                return;
+            }
+
+            var secondTypeRequired = ittype === "maplike";
+            var secondTypeAllowed = secondTypeRequired || ittype === "iterable";
+            ret.type = ittype;
+            if (ret.type !== 'maplike' && ret.type !== 'setlike')
+                delete ret.readonly;
+            all_ws();
+            if (consume(OTHER, "<")) {
```
Comment (marcoscaceres) [inline js/webidl2.js]: Cyclomatic complexity is way too high :( I'm worried about the maintainability of this code. @dontcallmedom, I know you said we should handle this stuff kind of code going forward.

**Comment by marcoscaceres** [comment]
LGTM, but we REALLY need to start adopting some better coding standards. This is nothing on @dontcallmedom, who does an amazing job following the current coding conventions used by ReSpec... but as a team, we need to change how we do ReSpec or we are just going to keep building up technical debt.

**Comment by halindrome** [comment]
Care to suggest some coding conventions and a tool for checking them?

On Thu, Dec 10, 2015 at 9:37 AM, Marcos Caceres <notifications@github.com>
wrote:

> LGTM, but we REALLY need to start adopting some better coding standards.
> This is nothing on @dontcallmedom <https://github.com/dontcallmedom>, who
> does an amazing job following the current coding conventions used by
> ReSpec... but as a team, we need to change how we do ReSpec or we are just
> going to keep building up technical debt.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/respec/pull/535#issuecomment-163661557>.
>



-- 
Shane McCarron
halindrome@gmail.com

**Comment by dontcallmedom** [comment]
while I sympathize with the need for better coding conventions (and ideally, Travis-CI enforced ones), the changes in webidl2.js should *really* be ignored given this is an import of an external library.

**Comment by tobie** [comment]
> while I sympathize with the need for better coding conventions […] the changes in webidl2.js should _really_ be ignored given this is an import of an external library.

Yup

**Comment by marcoscaceres** [comment]
yep, that's fine. Thanks for the info @dontcallmedom!

**Pull request merged**